// @sanity/base/lib/datastores/utils/createActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createActions;

var _nanoPubsub = _interopRequireDefault(require("nano-pubsub"));

var _rxjs = require("rxjs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Todo: refactor out
function createActions(actions) {
  return Object.keys(actions).reduce((acc, name) => {
    acc[name] = createAction(name, actions[name]);
    return acc;
  }, {});
}

function createAction(name, fn) {
  const calls = (0, _nanoPubsub.default)();

  const functor = function action() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    // eslint-disable-line func-name-matching
    const retValue = fn(...args);
    calls.publish({
      name: name,
      progress: retValue.progress,
      returnValue: retValue,
      args: args
    });
    return retValue;
  };

  functor.calls = new _rxjs.Observable(observer => {
    return calls.subscribe(call => {
      observer.next(call);
    });
  });
  return functor;
}
